### PR TITLE
Add option to parmout and chamber to write VMD-compatible prmtop files

### PR DESCRIPTION
**Pull request by Eigenstate**
VMD has a very strict interpretation of the prmtop format, and ignores the format fields when reading one in. This means that bond data is not successfully parsed from prmtop files that parmed outputs, preventing visualization of the results.

I added an option, -vmd to chamber and parmout that will result in the format fields of the prmtop being what VMD expects. This has the same function as the -vmd flag to standalone chamber.

Example usage:
action = parmout(parm, "test.prmtop test.inpcrd vmd")

### CODE REVIEW

File: chemistry/amber/amberformat.py [hunk 1]
```diff
@@ -814,15 +834,24 @@ def write_parm(self, name):
             for i in xrange(len(self.parm_data[self.charge_flag])):
                 self.parm_data[self.charge_flag][i] *= CHARGE_SCALE
 
+        if self.vmd_compat is not None:
+            self.delete_flag('CTITLE')
```
Comment (swails) [inline chemistry/amber/amberformat.py]: This should be moved into `convert_to_vmd_compat` -- `write_parm` shouldn't modify the parm object (permanently at least).

File: chemistry/amber/amberformat.py [hunk 2]
```diff
@@ -814,15 +834,24 @@ def write_parm(self, name):
...
         # write data to prmtop file, inserting blank line if it's an empty field
         for flag in self.flag_list:
             new_prm.write('%%FLAG %s\n' % flag)
             # Insert any comments before the %FORMAT specifier
-            for comment in self.parm_comments[flag]:
-                new_prm.write('%%COMMENT %s\n' % comment)
+            # except VMD hates comments
+            if self.vmd_compat is None:
+              for comment in self.parm_comments[flag]:
+                  new_prm.write('%%COMMENT %s\n' % comment)
```
Comment (swails) [inline chemistry/amber/amberformat.py]: Let's just have `convert_to_vmd_compat` delete all comments and get rid of the `vmd_compat` flag altogether.

File: chemistry/amber/amberformat.py [hunk 3]
```diff
@@ -788,6 +787,27 @@ def set_version(self):
 
     #===================================================
 
+    def convert_to_vmd_compat(self):
```
Comment (swails) [inline chemistry/amber/amberformat.py]: `AmberFormat` is a base class to the actual topology file objects.  `AmberParm` inherits from `AmberFormat` and both `ChamberParm` and `AmoebaParm` inherit from `AmberParm`.  This method is specific to chamber topology files (and will not in fact work with any other flavor).  So can you move this function to the `ChamberParm` class in `chemistry/amber/_chamberparm.py`?

File: ParmedTools/ParmedActions.py [hunk 4]
```diff
@@ -261,6 +265,8 @@ def execute(self):
         if self.rst_name is not None:
             if not Action.overwrite and os.path.exists(self.rst_name):
                 raise FileExists('%s exists; not overwriting.' % self.rst_name)
+        if self.vmd is not None:
+            self.parm.convert_to_vmd_compat()
```
Comment (swails) [inline ParmedTools/ParmedActions.py]: I don't think that `parmout` should modify the topology file -- it goes against what I would expect to happen.  So if you want the `vmd` option in `parmout` (and `outparm`), perhaps make a copy of `self.parm` and convert that to VMD compatibility... but this probably won't support `AmoebaParm` correctly yet...

**Comment by swails** [comment]
The main reason I didn't add this capability when I first implemented the action is because I figured people should just use the PSF file with VMD when visualizing, and a VMD-compatible chamber prmtop would run in Amber and just give *wrong* results -- but not obviously-wrong-or-crash results.

But then I ran into a case from CHARMM-GUI where cross-linked molecules needed to be rearranged in order to work with the `ATOMS_PER_MOLECULE` and `SOLVENT_POINTERS` sections of the prmtop, so the PSF file won't have the same atom ordering as the resulting topology file, so this is a good addition.

A couple questions/comments:

* The `vmd_compat` flag is causing problems.  This attribute doesn't exist unless the `convert_to_vmd_compat` function is called, but it's used in `write_parm` (which means that `write_parm` fails with an `AttributeError` any time you call it on a prmtop file that hasn't been passed to `convert_to_vmd_compat` (that's why the Travis CI tests are failing).  It would be best if `chamber` just took this option and called `convert_to_vmd_compat` and made all of the necessary changes to implement this change directly in that function (rather than offloading some of the work to `write_parm`).
* `convert_to_vmd_compat` is specific to chamber topologies, so it should only be an instance method on `ChamberParm`.
* What, if anything, should we do to "mutilate" the resulting VMD-compatible prmtop so it's valid for visualization but will crash when simulating (make the bond force constants all incredibly large, for instance)?

BTW -- nice to see you on Github!  I hope you're enjoying Stanford

**Comment by Eigenstate** [comment]
Hmmmm. I was actually not aware that the VMD-compatible prmtops aren't read in correctly by sander/pmemd, which is kind of concerning.

Since I've been messing with the VMD code a lot lately, I think I will just fix their prmtop parser so this sort of workaround isn't necessary, and you can table this pull request until then.

Grad school is great! Also github is addictive. Hope you're well :)

**Comment by swails** [comment]
> Hmmmm. I was actually not aware that the VMD-compatible prmtops aren't read in correctly by sander/pmemd, which is kind of concerning.

Well they're read in "correctly" -- but they're read in as a "regular" prmtop file (which means the CHARMM-specific CMAP and improper torsion parameters) are ignored.  So it's "close enough" to a real force field that the simulation doesn't blow up, but obviously without the CMAPs and impropers, it's just wrong.  So there is no error message, the simulation doesn't fail, and the results are not obviously wrong -- the trifecta of "how not what to act like".

> Since I've been messing with the VMD code a lot lately, I think I will just fix their prmtop parser so this sort of workaround isn't necessary, and you can table this pull request until then.

Cool.  If it's of any use, ParmEd has a fairly generalized C++ prmtop parser (that's actually used to accelerate prmtop reading in Python).  The code's in `src/readparm.cpp`, and should be fairly robust against future prmtop extensions.  But if this PR would help you with your work, then I'm more than happy to get it merged in when it's fixed up a bit.

> Grad school is great! Also github is addictive. Hope you're well :)

Good to hear (and agreed -- Github is addictive :)).

**Comment by swails** [comment]
this is stale and unmergeable at the moment, so I'm going to close it.

**Pull request closed without merging**